Misc. fixes for Coverity defects

peter-b opened the pull request:
Coverity-Id: 120052
Coverity-Id: 26003
Coverity-Id: 16051

montegoulding commented on engine/src/srvcgi.cpp:
```diff
@@ -1816,6 +1816,9 @@ static bool cgi_send_headers(void)
 			case kMCSOutputTextEncodingUTF8:
 				sprintf(t_content_header, "Content-Type: text/html; charset=utf-8\n");
 				break;
+			default:
```
@peter-b is there a reason not to handle `kMCSOutputTextEncodingUndefined` here so when we turn on  `-Werror=switch` we we will get an error when we add an output encoding and don't handle it? I did handle this case in my switch patch but I didn't actually output a header as you are which seems sensible... or should this actually be an unreachable state?

montegoulding replied:
Actually I wonder if we can just remove `kMCSOutputTextEncodingUndefined` entirely... it doesn't appear there's a code pat that can give `MCserveroutputtextencoding` that value...

peter-b replied:
Sure, I'll look at both options. Good spot. 😎

peter-b committed:
Remove kMCSOutputTextEncodingUndefined

The output text encoding should always be defined -- it should not be
possible to set it to an invalid value.  Remove the
`kMCSOutputTextEncodingUndefined` constant entirely.

Coverity-Id: 16051

peter-b committed:
lcidl: Fix memory leak in constant parsing.

Coverity-Id: 120052

peter-b committed:
Define kMCSOutputTextEncodingNative outside MCSOutputTextEncoding enum

When `kMCSOutputTextEncodingNative` is defined within the
`MCSOutputTextEncoding`, it is not possible to write a `switch`
statement over all members of the enumeration because
`kMCSOutputTextEncodingNative` aliases with one of the other members.

Making `kMCSOutputTextEncodingNative` into a constant defined outside
the enumeration means that it can be used as a default value for a
text encoding, but allows a `switch` statement over
`MCSOutputTextEncoding` to be validated as complete.

Coverity-Id: 16051

peter-b committed:
MCIdeScriptClassify(): Throw if target object isn't available.

Modify `MCIdeScriptClassify()` to throw an error immediately if the
target object is not available.

Coverity-Id: 26003

peter-b committed:
libscript: Tweak MCScriptForeignInvocation to help Coverity

The `MCScriptForeignInvocation` implementation is confusing Coverity,
which is reporting a bunch of uninitialised memory errors relating to
pointers acquired via the `MCScriptForeignInvocation::Allocate()`
method.  There were two possible problems:

- The maximum storage allocatable was stored in
  `MCScriptForeignInvocation::m_storage_limit`, a mutable member
  variable (despite being a compile-time constant).  It's possible
  that Coverity was assuming that it could be mutated at some point to
  be larger than the size of the fixed-size `m_stack_storage` array,
  allowing `Allocate()` to return a pointer to invalid memory.  This
  patch removes the member.

- `MCScriptForeignInvocation::ReferenceArgument()` and `Argument()`
  were checking whether the current argument count was _equal_ to the
  maximum number of arguments, rather than checking that it was less
  than the maximum.  This could allow Coverity to assume that the
  current argument count was greater than the maximum number of
  arguments, once again allowing pointers to invalid memory to be
  produced.  This patch introduces inequalities.

Coverity-Id: 137925
Coverity-Id: 137927

peter-b committed:
libscript: Prevent null dereference on error when property has no getter

The `MCScriptGetTypeOfPropertyInModule()` function could deference a
null pointer when attempting to determine the get type for the
specified property.  This function was only used in the
`MCScriptThrowInvalidValueForPropertyError()` error dispatch helper,
which in turn was only used from contexts where the VM had already
resolved a specific type for the property and had _just_ failed to
compare a value with it.

This patch removes the `MCScriptGetTypeOfPropertyInModule()` function
entirely, and refactors `MCScriptThrowInvalidValueForPropertyError()`
to take the failed property type as an argument.

By doing so it also corrects another bug: when the type required when
setting a property is different from the type obtained by getting the
property, the error message when failing to set the property would be
misleading.

Coverity-Id: 137924

peter-b committed:
libscript: Ensure all members of MCScriptExecuteContext are initialised

Explicitly initialise the `m_operation` and `m_argument_count` members
of `MCScriptExecuteContext`.

This isn't exactly right, because there _isn't_ a suitable default
bytecode opcode.

Coverity-Id: 137929

peter-b committed:
libscript: Check "bytecode ready" MCScriptExecuteContext invariant

`MCScriptOperationContext` allows direct access to several of its
fields via accessor functions.  However, these fields only contain
valid values if the `Step()` method has previously been called
successfully.

This patch adds an `m_operation_ready` member to the class and adds
assertions to the accessor to ensure that they fail before the
required fields have actually been populated.

montegoulding commented:
restarting travis after an IDE test crash

peter-b committed:
libscript: Don't mix enum types.

The on-disk value saved in the compiled LCB bytecode for the mode of a
handler parameter (`in`/`out`/`inout`) is described by the
`MCScriptHandlerTypeParameterMode` enumeration.  However, when
constructing an `MCHandlerTypeInfo` value for use at runtime, these
values need to be translated to the `MCHandlerTypeFieldMode`
enumeration.  This patch adds an explicit `inline` conversion method
to the `MCScriptHandlerTypeParameter` structure.

Coverity-Id: 16984

peter-b committed:
Add missing semicolon.

A semicolon was missed out from commit ae3097b7.

livecodefraser commented:
@livecode-vulcan review okay 5d59fd3

livecode-vulcan commented:
💙 review by @livecodefraser ok 5d59fd352dd88cf5d9f65e4736a5871537761ae2

livecodeali commented on engine/src/osspec.h:
```diff
@@ -240,23 +240,21 @@ MCSErrorMode MCS_get_errormode(void);
+
 #if defined(__MACROMAN__)
-	kMCSOutputTextEncodingNative = kMCSOutputTextEncodingMacRoman,
+static const MCSOutputTextEncoding kMCSOutputTextEncodingNative = kMCSOutputTextEncodingMacRoman
```
There's a missing semicolon here.

livecode-vulcan commented:
😞 test failure 5d59fd352dd88cf5d9f65e4736a5871537761ae2
- try-community-linux-x86: [pending](https://vulcan.livecode.com/builders/try-community-linux-x86/builds/489)
- try-community-mac: [failure](https://vulcan.livecode.com/builders/try-community-mac/builds/494)
- try-community-linux-x86_64: [pending](https://vulcan.livecode.com/builders/try-community-linux-x86_64/builds/489)
- try-community-emscripten: [pending](https://vulcan.livecode.com/builders/try-community-emscripten/builds/489)
- try-community-ios: [failure](https://vulcan.livecode.com/builders/try-community-ios/builds/495)
- try-community-win-x86: [pending](https://vulcan.livecode.com/builders/try-community-win-x86/builds/489)
- try-community-android-armv6: [pending](https://vulcan.livecode.com/builders/try-community-android-armv6/builds/489)

peter-b commented:
@livecodeali Thanks -- I've fixed that!

livecodefraser commented:
@livecode-vulcan review okay fa2e223

livecode-vulcan commented:
💙 review by @livecodefraser ok fa2e22387537898b611d4dfcef121c722b582f0c

livecode-vulcan commented:
😎 test success fa2e22387537898b611d4dfcef121c722b582f0c
- try-community-linux-x86: [success](https://vulcan.livecode.com/builders/try-community-linux-x86/builds/491)
- try-community-mac: [success](https://vulcan.livecode.com/builders/try-community-mac/builds/496)
- try-community-linux-x86_64: [success](https://vulcan.livecode.com/builders/try-community-linux-x86_64/builds/491)
- try-community-emscripten: [success](https://vulcan.livecode.com/builders/try-community-emscripten/builds/491)
- try-community-ios: [success](https://vulcan.livecode.com/builders/try-community-ios/builds/497)
- try-community-win-x86: [success](https://vulcan.livecode.com/builders/try-community-win-x86/builds/491)
- try-community-android-armv6: [success](https://vulcan.livecode.com/builders/try-community-android-armv6/builds/491)

peter-b commented:
@livecode-vulcan retry fa2e223

livecode-vulcan commented:
😎 test success fa2e22387537898b611d4dfcef121c722b582f0c
- try-community-linux-x86: [success](https://vulcan.livecode.com/builders/try-community-linux-x86/builds/492)
- try-community-mac: [success](https://vulcan.livecode.com/builders/try-community-mac/builds/497)
- try-community-linux-x86_64: [success](https://vulcan.livecode.com/builders/try-community-linux-x86_64/builds/492)
- try-community-emscripten: [success](https://vulcan.livecode.com/builders/try-community-emscripten/builds/492)
- try-community-ios: [success](https://vulcan.livecode.com/builders/try-community-ios/builds/498)
- try-community-win-x86: [success](https://vulcan.livecode.com/builders/try-community-win-x86/builds/492)
- try-community-android-armv6: [success](https://vulcan.livecode.com/builders/try-community-android-armv6/builds/492)
